core/trino-main/src/main/java/io/trino/server/ServerMainModule.java

```diff
@@ -271,6 +280,22 @@ protected void setup(Binder binder)
         binder.bind(NodeTaskMap.class).in(Scopes.SINGLETON);
         newExporter(binder).export(NodeScheduler.class).withGeneratedName();
 
+        // node allocator
+        install(conditionalModule(
+                NodeSchedulerConfig.class,
+                config -> FIXED_COUNT == config.getNodeAllocatorType(),
+                innerBinder -> {
+                    innerBinder.bind(NodeAllocatorService.class).to(FixedCountNodeAllocatorService.class).in(Scopes.SINGLETON);
+                    innerBinder.bind(PartitionMemoryEstimator.class).to(ConstantPartitionMemoryEstimator.class).in(Scopes.SINGLETON);
+                }));
+        install(conditionalModule(
+                NodeSchedulerConfig.class,
+                config -> FULL_NODE_CAPABLE == config.getNodeAllocatorType(),
```

Comment:
Actually I want to add proper BIN_PACKING implementation as a separate one. Eventually we can drop some of those - when we see that one is strictly better than the other. But initially I would rather have some options to choose from, so we can experiment what works best.

Reply:
The current implementation is already a "bin packing" implementation with a fallback to a full node allowed. The improvements to it could be iterative. For example it should be pretty straightforward to change the logic (by adjusting the `PartitionMemoryEstimator`) to first retry with, say double the memory and only then to fallback to a "full node". Thus it doesn't feel like a new implementation of the `NodeAllocator` would be needed. Thoughts?

Reply:
Actually, I think the implementation will quite different. I will not have special handling for "full-node". And given fact that I can get different memory requirements I will need to be much more careful about the liveliness problem. It will be much easier to starve a request. 
Eventually we want to have just one implementation, but for time being I think it is safer to always have something fairly tested, to fallback to, while working on new thing, even if this new thing is natural extension what we currently have.